### PR TITLE
net: sockets: dtls: Fix handshake with socket offloading

### LINKED ISSUE #35789: sockets_tls: receiving data on offloaded socket before handshake causes pollin | pollerr and failed recvfrom (SARA-R4)
**Describe the bug**
This bug is a variant of https://github.com/zephyrproject-rtos/zephyr/issues/33330 but with offloaded socket behind the DTLS socket. The fix for mentioned issue doesn't apply for offloaded sockets.

The scenario is as follows:
1. A DTLS socket is created and `bind()` is called with an offloaded socket behind it 
2. A packet comes on the offloaded socket before any other calls has been made (see picture, SOCR->bind() UUSORF->data waiting on socket)
bugsara
3. when poll() is called, return code is 1 with revents=9 (POLLIN | POLLERR).
4. Since POLLIN is set, recvfrom is called but since no data exist in mbedTLS it returns a failure

The reason as to why mentioned fix doesn't work for offloaded sockets appears to be that for *native* sockets ioctl is implemented while sara-r4 modem's `offload_ioctl` just returns `-EXDEV` on `POLL_PREPARE` request, whereby we just restore the events and go on like nothing happened (aka. patch is ignored).
https://github.com/zephyrproject-rtos/zephyr/blob/8e6898ef14560c7f20b35f452fb4fec843f39728/subsys/net/lib/sockets/sockets_tls.c#L1981-L2022
**To Reproduce**
1. Hard to reproduce, best inject a `UUSORF` just after creating the socket to simulate, for example using a debugger.

**Expected behavior**
`poll()` on DTLS socket where handshake not complete should always return nothing.

**Impact**
Possible network hang in rare cases when network is inited and deinited often. I guess this is not super serious, but I've seen it happen a few times during our unit tests when we stress-test it by bringing our network stack up&down to see its behavior. *All special cases are not handled*

**Logs and console output**
See above

**Environment (please complete the following information):**
 - STM32H747I DISCO devkit with Sparkfun SARA-R4 shield
 - zephyr 9867ac2a

### PR DESCRIPTION
**Pull request by rlubos**
Fix `poll()` handling for DTLS clients when the underlying socket is an
offloaded socket. As in this case no `k_poll()` is used underneath, it's
not possible to monitor the handhshake status with `tls_established`
semaphore. Instead, do the following:

1. If no handhshake is in progress yet, just drop the incoming data -
   it's the client who should initiate the handshake, any data incoming
   before that should not be processed.
2. If handshake is currently in progress, lift the `POLLIN` flag and add
   small delay to allow the other thread to proceed with the handshake.
3. Otherwise, just proceed as usual.

Fixes #35789

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

### CODE REVIEW

**Comment by rlubos** [comment]
CC @emillindq

**Comment by emillindq** [comment]
@rlubos Firstly sorry for the delay!

I've tried this PR out with the following test:
* Ublox SARA-R4 modem
* Target (aka. my devkit running Zephyr) creates DTLS socket and binds, and then `sendto()` to server
* On server is a simple script which is listening on UDP port and then spams the crap out of the target with UDP messages whenever it sends something
* Handshake will fail and it will be spammed on failed handshake
* After a while target does close, socket, bind, and then `poll()`. The server is still spamming and the NAT is open on the LTE network -> UDP datagrams will come to SARA-R4 modem right when `socket()` and `bind()` is called the second time.

With PR~1, same behavior as before. With PR commit, the waiting UDP packet is successfully discarded if handshake is not started. Thanks for this! :+1:

**Comment by rlubos** [comment]
@emillindq Thanks for verifying. So, shall I pull the PR out of the draft stage then?

**Comment by emillindq** [comment]
> @emillindq Thanks for verifying. So, shall I pull the PR out of the draft stage then?

I'm just thinking about the name `poll_offload_dtls_client_check` should be more descriptive. What is it returning? I'd expect to get an understanding of what the function does by looking at the name. If it does multiple things, then maybe it should be split up? Just my humble opinion :smiley_cat:

**Comment by rlubos** [comment]
> > @emillindq Thanks for verifying. So, shall I pull the PR out of the draft stage then?
> 
> I'm just thinking about the name `poll_offload_dtls_client_check` should be more descriptive. What is it returning? I'd expect to get an understanding of what the function does by looking at the name. If it does multiple things, then maybe it should be split up? Just my humble opinion 😺

It basically verifies if the `poll()` should be repeated for DTLS client. How about `poll_offload_dtls_client_retry()` then? 
`if (poll_offload_dtls_client_retry())` is quite verbose to me.

**Comment by emillindq** [comment]
> > > @emillindq Thanks for verifying. So, shall I pull the PR out of the draft stage then?
> > 
> > 
> > I'm just thinking about the name `poll_offload_dtls_client_check` should be more descriptive. What is it returning? I'd expect to get an understanding of what the function does by looking at the name. If it does multiple things, then maybe it should be split up? Just my humble opinion smiley_cat
> 
> It basically verifies if the `poll()` should be repeated for DTLS client. How about `poll_offload_dtls_client_retry()` then?
> `if (poll_offload_dtls_client_retry())` is quite verbose to me.

Yes great! :+1:

**Comment by rlubos** [comment]
@emillindq Thanks for all the feedback, I've renamed the funcion. Feel free to approve if you have no more concerns :)

**Comment by emillindq** [comment]
I'm not a reviewer so I can't approve the PR, but you have my blessings :smile:

**Comment by rlubos** [comment]
> I'm not a reviewer so I can't approve the PR, but you have my blessings 😄

You don't need to be a "requested" reviewer to drop a review. Sorry, but I wasn't able to add you to the list :/

**Review by mniestroj — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**
